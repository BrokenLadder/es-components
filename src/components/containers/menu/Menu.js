import React from 'react';
import PropTypes from 'prop-types';
import ToggleButton from '../../controls/buttons/ToggleButton';
import styled, { ThemeProvider } from 'styled-components';
import MenuPanel from './MenuPanel';
import MenuSection from './MenuSection';
import RootCloseWrapper from 'react-overlays/lib/RootCloseWrapper';
import defaultTheme from '../../theme/defaultTheme';

const Backdrop = styled.div`
  background-color: black;
  bottom: 0;
  cursor: auto;
  left: 0;
  opacity: 0.5;
  position: fixed;
  right: 0;
  top: 0;
  z-index: ${props => (props.isMenuOpen ? 'auto' : '-1')};
`;

class Menu extends React.Component {
  static childContextTypes = {
    inline: PropTypes.bool
  };

  state = {
    isMenuOpen: false
  };

  getChildContext() {
    return { inline: this.props.inline };
  }

  toggleMenu = () => {
    this.setState(previousState => ({ isMenuOpen: !previousState.isMenuOpen }));
  };

  closeMenu = () => {
    if (this.state.isMenuOpen) {
      this.setState({ isMenuOpen: false });
    }
  };

  render() {
    const {
      children,
      buttonContent,
      className,
      theme,
      openButtonType,
      rootClose,
      hasBackdrop
    } = this.props;

    return (
      <ThemeProvider theme={theme}>
        <RootCloseWrapper onRootClose={this.closeMenu} disabled={!rootClose}>
          <div className={className}>
            {hasBackdrop && (
              <Backdrop
                isMenuOpen={this.state.isMenuOpen}
                onClick={this.closeMenu}
              />
            )}
            <ToggleButton
              handleOnClick={this.toggleMenu}
              isPressed={this.state.isMenuOpen}
              styleType={openButtonType}
            >
              {buttonContent}
            </ToggleButton>
            <MenuPanel isOpen={this.state.isMenuOpen} onClose={this.closeMenu}>
              {children}
            </MenuPanel>
          </div>
        </RootCloseWrapper>
      </ThemeProvider>
    );
  }
}

Menu.MenuSection = MenuSection;

Menu.propTypes = {
  children: PropTypes.any.isRequired,
  buttonContent: PropTypes.any.isRequired,
  openButtonType: PropTypes.string,
  className: PropTypes.string,
  rootClose: PropTypes.bool,
  inline: PropTypes.bool,
  theme: PropTypes.object,
  hasBackdrop: PropTypes.bool
};

Menu.defaultProps = {
  theme: defaultTheme,
  rootClose: false,
  openButtonType: 'default'
};

export default Menu;
