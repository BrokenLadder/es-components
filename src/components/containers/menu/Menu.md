Use the `Menu` to expose an list of items overlayed over the screen when a button is pressed. All nodes should be `Menu.MenuSection`s

### Basic Example
```
<Menu header="Small Menu" buttonContent="Open Menu">
  <Menu.MenuSection title="Menu Section" isFirst>
    <a href="www.google.com">Go To Google</a>
  </Menu.MenuSection>
  <Menu.MenuSection title="Menu Section 2">
    <a href="www.yahoo.com">Go To Yahoo</a>
  </Menu.MenuSection>
  <Menu.MenuSection title="Menu Section 3">
    <a href="www.lycos.com">Go To Lycos</a>
  </Menu.MenuSection>
  <Menu.MenuSection title="Menu Section 4">
    <a href="www.dogpile.com">Go To Dogpile</a>
  </Menu.MenuSection>
  <Menu.MenuSection title="Menu Section 5">
    <a href="www.bing.com">Go To Bing</a>
  </Menu.MenuSection>
  <Menu.MenuSection title="Menu Section 6">
    <a href="www.ask.com">Go To Ask</a>
  </Menu.MenuSection>
  <Menu.MenuSection title="Menu Section 7">
    <a href="www.search.com">Go To Search</a>
  </Menu.MenuSection>
  <Menu.MenuSection title="Menu Section 8" isLast>
    <a href="www.duckduckgo.com">Go To Duck Duck Go</a>
  </Menu.MenuSection>
</Menu>
```