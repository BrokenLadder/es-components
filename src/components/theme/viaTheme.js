const white = '#fff';
const gray1 = '#f4f4f4';
const gray2 = '#ecedee';
const gray3 = '#e4e4e4';
const gray4 = '#d8d8d8';
const gray5 = '#c5c5c5';
const gray6 = '#979797';
// const gray7 = '#707070';
const gray8 = '#585858';
const gray9 = '#444';
const black = '#000';

const grayLightest = gray1;
const grayLighter = gray2;
const grayLight = gray3;
const gray = gray5;
const grayDark = gray6;
const grayDarker = gray8;
const grayDarkest = gray9;

const dflt = '#d8d8d8';
const dfltHover = '#bfbfbf';
const primary = '#007fa7';
const primaryHover = '#005874';
const info = '#069';
const infoHover = '#004466';
const success = '#298544';
const successHover = '#1d5e30';
const warning = '#b35f00';
const warningHover = '#804400';
const danger = '#c00';
const dangerHover = '#990000';
const advisor = warning;

const boxShadowLight = 'rgba(0, 0, 0, 0.075)';
const boxShadowDark = 'rgba(0, 0, 0, 0.5)';

const inputBorder = gray4;
const inputFocus = '#83bffc';
const inputBoxShadow = 'rgba(102, 175, 233, 0.6)';

// Datepicker colors
const navArrow = gray;
const navArrowHover = grayDark;
const selected = '#216ba5';
const hover = '#1d5d90';
const keyboard = '#2a87d0';
const inRange = 'rgba(33, 107, 165, 0.5)';
const highlight = '#3dcc4a';
const highlightHover = '#32be3f';

const theme = {
  colors: {
    black,
    grayDarkest,
    grayDarker,
    grayDark,
    gray,
    grayLight,
    grayLighter,
    grayLightest,
    white,
    dflt,
    dfltHover,
    primary,
    primaryHover,
    success,
    successHover,
    info,
    infoHover,
    warning,
    warningHover,
    danger,
    dangerHover,
    advisor,
    boxShadowLight,
    boxShadowDark,
    inputFocus
  },
  datepickerColors: {
    navArrow,
    navArrowHover,
    selected,
    hover,
    keyboard,
    inRange,
    highlight,
    highlightHover
  },
  buttonStyles: {
    buttonsNormal: {
      primary: {
        bgColor: primary,
        textColor: white,
        hoverBgColor: primaryHover,
        hoverTextColor: white,
        activeBgColor: primaryHover,
        activeTextColor: white,
        boxShadowColor: primaryHover
      },
      secondary: {
        bgColor: dflt,
        textColor: white,
        hoverBgColor: dfltHover,
        hoverTextColor: white,
        activeBgColor: dfltHover,
        activeTextColor: white,
        boxShadowColor: dfltHover
      },
      success: {
        bgColor: success,
        textColor: white,
        hoverBgColor: successHover,
        hoverTextColor: white,
        activeBgColor: successHover,
        activeTextColor: white,
        boxShadowColor: successHover
      },
      info: {
        bgColor: info,
        textColor: white,
        hoverBgColor: infoHover,
        hoverTextColor: white,
        activeBgColor: infoHover,
        activeTextColor: white,
        boxShadowColor: infoHover
      },
      warning: {
        bgColor: warning,
        textColor: white,
        hoverBgColor: warningHover,
        hoverTextColor: white,
        activeBgColor: warningHover,
        activeTextColor: white,
        boxShadowColor: warningHover
      },
      danger: {
        bgColor: danger,
        textColor: white,
        hoverBgColor: dangerHover,
        hoverTextColor: white,
        activeBgColor: dangerHover,
        activeTextColor: white,
        boxShadowColor: dangerHover
      }
    },
    buttonsAlt: {
      primary: {
        bgColor: white,
        textColor: primary,
        hoverBgColor: primary,
        hoverTextColor: white,
        activeBgColor: primaryHover,
        activeTextColor: white,
        borderColor: primary
      },
      gray: {
        bgColor: white,
        textColor: '#63666a',
        hoverBgColor: '#63666a',
        hoverTextColor: white,
        activeBgColor: '#4a4d50',
        activeTextColor: white,
        borderColor: '#63666a'
      },
      magenta: {
        bgColor: white,
        textColor: '#c110a0',
        hoverBgColor: '#c110a0',
        hoverTextColor: white,
        activeBgColor: '#920C79',
        activeTextColor: white,
        borderColor: '#c110a0'
      },
      violet: {
        bgColor: white,
        textColor: '#702082',
        hoverBgColor: '#702082',
        hoverTextColor: white,
        activeBgColor: '#4d1659',
        activeTextColor: white,
        borderColor: '#702082'
      }
    }
  },
  // buttonSizes should always have default, lg, sm, xs
  buttonSizes: {
    lg: {
      borderRadius: '3px',
      fontSize: '26px',
      padding: '8px 20px 5px'
    },
    default: {
      borderRadius: '2px',
      fontSize: '18px',
      padding: '5px 15px 4px'
    },
    sm: {
      borderRadius: '2px',
      fontSize: '14px',
      padding: '4px 10px 3px'
    },
    xs: {
      borderRadius: '2px',
      fontSize: '12px',
      padding: '3px 10px 2px'
    }
  },
  validationIconName: {
    success: 'ok',
    warning: 'warning-sign',
    danger: 'remove'
  },
  validationTextColor: {
    success,
    warning: warningHover,
    danger
  },
  validationInputColor: {
    success: {
      borderColor: success,
      boxShadow: `inset 0 0 5px ${success}`,
      focusBorderColor: success,
      focusBoxShadow: `0 0 6px ${success}`
    },
    warning: {
      borderColor: warning,
      boxShadow: `inset 0 0 5px ${warning}`,
      focusBorderColor: warning,
      focusBoxShadow: `0 0 6px ${warning}`
    },
    danger: {
      borderColor: danger,
      boxShadow: `inset 0 0 5px ${danger}`,
      focusBorderColor: danger,
      focusBoxShadow: `0 0 6px ${danger}`
    },
    default: {
      borderColor: inputBorder,
      boxShadow: `inset 0 1px 1px ${boxShadowLight}`,
      focusBorderColor: inputFocus,
      focusBoxShadow: `0 0 8px ${inputBoxShadow}`
    }
  },
  sizes: {
    baseFontSize: 18,
    baseLineHeight: 24
  },
  screenSize: {
    retina: '320px',
    phone: '480px',
    tablet: '768px',
    desktop: '992px',
    widescreen: '1200px'
  }
};

export default theme;
