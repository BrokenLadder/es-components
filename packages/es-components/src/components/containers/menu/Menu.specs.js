/* eslint-env jest */

import React from 'react';
import { render, fireEvent } from 'react-testing-library';
import { ThemeProvider } from 'styled-components';
import viaTheme from 'es-components-via-theme';
import { Menu } from './Menu';

function buildMenu() {
  return (
    <ThemeProvider theme={viaTheme}>
      <Menu
        headerContent="Small Menu"
        buttonContent="Open Menu"
        className="test"
      >
        <Menu.MenuSection title="Menu Section" isFirst>
          <a href="#test">Test link</a>
        </Menu.MenuSection>
      </Menu>
    </ThemeProvider>
  );
}

it('toggles the menu open and closed', () => {
  const { getByText } = render(buildMenu());

  const menuToggleButton = getByText('Open Menu');
  const menuSection = getByText('Menu Section');

  menuToggleButton.click();
  expect(menuSection).toBeVisible();

  menuToggleButton.click();
  expect(menuSection).not.toBeVisible();
});

it('closes open menu when ESC is pressed', () => {
  const { container, getByText } = render(buildMenu());

  getByText('Open Menu').click();

  fireEvent.keyDown(container, { keyCode: 27 });
  expect(getByText('Menu Section')).not.toBeVisible();
});
