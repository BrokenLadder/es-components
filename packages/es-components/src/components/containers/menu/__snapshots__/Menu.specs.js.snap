// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MenuTestSuite renders as expected 1`] = `
<div
  className="es-menu test"
>
  <button
    aria-expanded={false}
    className="es-menu__toggle-button sc-ifAKCX drovqE sc-bwzfXH hJmlCu sc-bdVaJa kHkNOX"
    onClick={[Function]}
  >
    Open Menu
  </button>
  <div
    className="es-menu__panel sc-gzVnrw ZXtoB"
  >
    <header
      className="sc-dnqmqq jBxXuW"
    >
      <span>
        Small Menu
      </span>
      <button
        aria-label="Close"
        className="sc-htoDjs XrUxb sc-EHOje jHwaMP"
        onClick={[Function]}
      >
        <span
          aria-hidden="true"
        >
          Ã—
        </span>
        <span
          className="sc-bZQynM eBGzNP"
        >
          Close
        </span>
      </button>
    </header>
    <div
      className="sc-iwsKbI kyjaoD"
    >
      <section
        className="es-menu__section sc-gZMcBi ivZJCw"
      >
        <header
          aria-label="Menu Section"
          className="sc-gqjmRU fiUekS"
        >
          Menu Section
        </header>
        <section
          className="sc-VigVT frPBLq"
        >
          <a
            href="www.google.com"
          >
            Go To Google
          </a>
        </section>
      </section>
    </div>
  </div>
</div>
`;
